dvc/config.py

```diff
@@ -300,250 +261,134 @@ def init(dvc_dir):
         open(config_file, "w+").close()
         return Config(dvc_dir)
 
-    def _resolve_cache_path(self, config):
-        cache = config.get(self.SECTION_CACHE)
-        if cache is None:
-            return
-
-        cache_dir = cache.get(self.SECTION_CACHE_DIR)
-        if cache_dir is None:
-            return
-
-        cache[self.PRIVATE_CWD] = os.path.dirname(config.filename)
-
-    def _resolve_paths(self, config):
-        if config.filename is None:
-            return config
-
-        ret = copy.deepcopy(config)
-        self._resolve_cache_path(ret)
-
-        for section in ret.values():
-            if self.SECTION_REMOTE_URL not in section.keys():
-                continue
-
-            section[self.PRIVATE_CWD] = os.path.dirname(ret.filename)
-
-        return ret
-
-    def _load_configs(self):
-        system_config_file = os.path.join(
-            self.get_system_config_dir(), self.CONFIG
-        )
+    def load(self, validate=True):
+        """Loads config from all the config files.
 
-        global_config_file = os.path.join(
-            self.get_global_config_dir(), self.CONFIG
-        )
+        Raises:
+            dvc.config.ConfigError: thrown if config has invalid format.
+        """
+        conf = {}
+        for level in self.LEVELS:
+            if level in self.files:
+                _merge(conf, self.load_one(level))
 
-        self._system_config = configobj.ConfigObj(system_config_file)
-        self._global_config = configobj.ConfigObj(global_config_file)
-        self._repo_config = configobj.ConfigObj()
-        self._local_config = configobj.ConfigObj()
+        if validate:
+            conf = self.validate(conf)
 
-        if not self.dvc_dir:
-            return
+        self.clear()
+        self.update(conf)
 
-        config_file = os.path.join(self.dvc_dir, self.CONFIG)
-        config_local_file = os.path.join(self.dvc_dir, self.CONFIG_LOCAL)
+        # Add resolved default cache.dir
+        if not self["cache"].get("dir"):
+            self["cache"]["dir"] = os.path.join(self.dvc_dir, "cache")
 
-        self._repo_config = configobj.ConfigObj(config_file)
-        self._local_config = configobj.ConfigObj(config_local_file)
+    def load_one(self, level):
+        conf = _load_config(self.files[level])
+        conf = self._load_paths(conf, self.files[level])
 
-    @property
-    def config_local_file(self):
-        return self._local_config.filename
+        # Autovivify sections
+        for key in COMPILED_SCHEMA.schema:
+            conf.setdefault(key, {})
 
-    @property
-    def config_file(self):
-        return self._repo_config.filename
+        return conf
 
-    def load(self):
-        """Loads config from all the config files.
+    @staticmethod
+    def _load_paths(conf, filename):
+        def resolve(path):
+            if os.path.isabs(path) or re.match(r"\w+://", path):
+                return path
+            return RelPath(os.path.join(cwd, path))
```

Comment:
cwd is defined after this helper in which it is used. This is error-prone.

Reply:
Why?

Reply:
This is minor. Nothing functionally wrong about it because of the same namespace, but causes a mental hickup.

Reply:
Defining subfunction in the middle causes mental hiccup too)

Reply:
I can move it if that worries you.